src/cms/logic.py

```diff
@@ -0,0 +1,323 @@
+    except AttributeError:
+        return html.title.get_text().strip()
+
+
+def get_body(html):
+    body = html.find('body')
+    if not isinstance(body, Tag):
+        return
+    for non_content_selector in ['script', '.djdt-hidden']:
+        for element in body.select(non_content_selector):
+            element.decompose()
+    for element in body.find_all(text=lambda text: isinstance(text, Comment)):
+        element.extract()
+    return body
+
+
+def get_base():
+    press = press_models.Press.objects.first()
+    return press.site_url()
+
+
+def excluded_urls():
+    return [
+        journal.site_url() for journal in journal_models.Journal.objects.all()
+    ] + [
+        repo.site_url() for repo in repository_models.Repository.objects.all()
+    ]
+
+
+def url_in_scope(deeper_url):
+    base = get_base()
+    if urlparse(deeper_url).hostname != urlparse(base).hostname:
+        return False
+    for excluded_url in excluded_urls():
+        if excluded_url in deeper_url:
+            return False
+    return True
+
+
+def remove_fragment(url):
+    fragment = urlparse(url).fragment
+    return url.replace('#' + fragment, '')
+
+
+def url_is_unique(deeper_url):
+    def normalize(url):
+        url = remove_fragment(url)
+        if url.endswith('/'):
+            url = url[:-1]
+        return url
+
+    if deeper_url in fetched_urls:
+        return False
+    for fetched_url in fetched_urls:
+        if normalize(fetched_url) == normalize(deeper_url):
+            return False
+    return True
+
+
+def decompose_non_content_page_regions(body):
+    for non_content_selector in ['header', 'footer', 'h1']:
+        for element in body.select(non_content_selector):
+            element.decompose()
+
+
+def add_searchable_page(url):
+
+    html = get_page(url)
+    if not html:
+        return
+
+    body = get_body(html)
+    if not body:
+        return
+
+    for anchor in body.find_all('a'):
+        href = anchor.get('href', '').strip()
+        deeper_url = urljoin(url, href)
+        if url_in_scope(deeper_url) and url_is_unique(deeper_url):
+            add_searchable_page(deeper_url)
+
+    name = get_name(html)
+    decompose_non_content_page_regions(body)
+
+    add_searchable_page_parts(url, body)
+    add_search_index_document(url, name, body.get_text())
+
+
+def get_press_site_search_data():
+    """
+    Generates data for press-level site search index
+    to be used by MiniSearch
+    """
+
+    base = get_base()
+    add_searchable_page(base)
+
+    if not len(search_documents) > 0:
+        logger.error('Search data store is empty')
+
+    return search_documents
+
+
+def update_search_data(press_id=1):
+    press = press_models.Press.objects.get(pk=press_id)
+    docs_filename = os.path.join(
+        settings.SITE_SEARCH_DIR,
+        f'_press_{ press.pk }_documents.json'
+    )
+    docs_file, created = models.MediaFile.objects.get_or_create(
+        label=docs_filename
+    )
```

Comment:
It would be good to encapsulate this in a function, rather than repeat the logic in various functions: `get_or_create_search_file`